crates/hir-ty/src/consteval/tests.rs

```diff
@@ -1958,6 +1958,36 @@ fn const_generic_subst_fn() {
     );
 }
 
+#[test]
+fn layout_of_type_with_associated_type_field_defined_inside_body() {
+    check_number(
+        r#"
+trait Tr {
+    type Ty;
+}
+
+struct St<T: Tr>(T::Ty);
+
+const GOAL: i64 = {
+    // if we move `St2` out of body, the test will fail, as we don't see the impl anymore. That
+    // case will probably be rejected by rustc in some later edition, but we should support this
+    // case.
```

Comment:
Do you mean *only* moving out the `St2` struct definition (leaving the impl for `St2` in the block)? That's surprising that that does not work. I'm pretty sure it should ...

Reply:
Yes I mean keeping the impl inside of the block. You can reproduce that by hovering on goal. I guess the problem is that we don't pass the block of the goal, only the crate, so it can't normalize the associated type. But with this logic, this one shouldn't work either...